osu.Game/Rulesets/Difficulty/DifficultyCalculator.cs

```diff
@@ -171,9 +171,10 @@ static IEnumerable<Mod> createDifficultyAdjustmentModCombinations(ReadOnlyMemory
         /// Enumerates <see cref="DifficultyHitObject"/>s to be processed from <see cref="HitObject"/>s in the <see cref="IBeatmap"/>.
         /// </summary>
         /// <param name="beatmap">The <see cref="IBeatmap"/> providing the <see cref="HitObject"/>s to enumerate.</param>
+        /// <param name="mods">Mods to calculate objects with.</param>
         /// <param name="clockRate">The rate at which the gameplay clock is run at.</param>
         /// <returns>The enumerated <see cref="DifficultyHitObject"/>s.</returns>
-        protected abstract IEnumerable<DifficultyHitObject> CreateDifficultyHitObjects(IBeatmap beatmap, double clockRate);
+        protected abstract IEnumerable<DifficultyHitObject> CreateDifficultyHitObjects(IBeatmap beatmap, Mod[] mods, double clockRate);
```

Comment:
not sure that I agree with every single ruleset's difficulty calculator having to calculate this rate on their own manually. if the base difficulty calculator is providing this base structure already, then it also should provide the pre-baked clock rate.

one way to do this would be to pass a `Func<double, double>` delegate - or something of that sort - that would compute the track rate at a given original track time.

Reply:
the reason I went with this approach rather than something like you mentioned was in the event of a potential rate-changing mod that isn't as clear cut as windup or winddown, like for example maybe a sine wave clockrate thing, or perhaps a "randomizer" one that smoothly transitions between various clockrates every x seconds based off of a seed

Reply:
both of those examples given above are not incompatible with my suggestion - in both cases the track rate is a function of the original track time. which is what i'm suggesting to pass down here.

what *would* break my suggestion was e.g. if the track slowed down or sped up proportionally to the player's health, for example. i shudder to think how complicated accounting for something like that in diffcalc would be.